### ISSUE TITLE
COPY does not detect contents change when building from some .tar files

### ISSUE THREAD

**Issue by jerdfelt** [moby/moby#35403]
**Description**

When building an image using a .tar file, absolute paths in the .tar file will cause COPY to miss changes to the contet of files. This results in builds with differing content to have the same digest.

**Steps to reproduce the issue:**
1. docker build http://sventech.com/def1-problem.tar
2. docker build http://sventech.com/def2-problem.tar

**Describe the results you received:**
```
% docker build http://sventech.com/def1-problem.tar
Downloading build context from remote url: http://sventech.com/def1-problem.tar
Step 1/2 : FROM alpine:latest
 ---> 37eec16f1872
Step 2/2 : COPY test.txt /tmp/test.txt
 ---> Using cache
 ---> 7a48fa0f7018
Successfully built 7a48fa0f7018
% docker build http://sventech.com/def2-problem.tar
Downloading build context from remote url: http://sventech.com/def2-problem.tar
Step 1/2 : FROM alpine:latest
 ---> 37eec16f1872
Step 2/2 : COPY test.txt /tmp/test.txt
 ---> Using cache
 ---> 7a48fa0f7018
Successfully built 7a48fa0f7018

# Notice that both image builds have the same digest even though the contents of test.txt are different in each .tar file
```

**Describe the results you expected:**
I would expect different digests because the contents of test.txt are different in each .tar file.

**Additional information you deem important (e.g. issue happens only occasionally):**
I've also built def1-ok.tar and def2-ok.tar (at the same host as above) which have the identical contents to def[12]-problem.tar but the filenames don't have a leading /. The def[12]-ok.tar files work as expected and result in different images.

It appears that the leading / (absolute filename) is causing docker to get confused and not detect that the contents of the files are different during the COPY.

This caused a lot of confusion and troubleshooting trying to isolate why changes to files weren't being reflected in the result of `docker build`.

**Output of `docker version`:**
```
Client:
 Version:      17.10.0-ce
 API version:  1.33
 Go version:   go1.8.3
 Git commit:   f4ffd25
 Built:        Tue Oct 17 19:05:55 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.10.0-ce
 API version:  1.33 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   f4ffd25
 Built:        Tue Oct 17 19:07:28 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 17.10.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 0351df1c5a66838d0c392b4ac4cf9450de844e2d
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.11.8-300.fc26.x86_64
Operating System: Fedora 26 (Cloud Edition)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.858GiB
Name: docker-test.localdomain
ID: 4ILD:4BFL:DHFZ:BYZA:3LO3:SOQJ:7NPZ:I5NF:FOFG:4CCY:YEW6:S4XU
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
ping @tonistiigi

**Comment by ruudboon** [comment]
Seeing the same issues here on CoreOS

**Comment by thaJeztah** [comment]
ping @tonistiigi is this the same issue as https://github.com/moby/moby/issues/36122 ?

**Comment by tonistiigi** [comment]
@thaJeztah Yes, I think so. These tars have leading slashes in paths. We should probably have a fix to mitigate this by just cleaning the path that would fix this and #36122 , although as long as we use tars for cache checksum there will always be a possibility to have these weird cases.
